pkg/kubernetes/deployment/extnv1beta1/v1alpha1/rollout_status.go

```diff
@@ -0,0 +1,151 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// rolloutStatus  is a typed function that abstracts status message formation logic
+type rolloutStatus func(*deploy) string
+
+// rolloutStatuses contains a group of status message for each predicate checks.
+// It useses predicateName as key.
+var rolloutStatuses = map[predicateName]rolloutStatus{
+	// PredicateProgressDeadlineExceeded refer to rolloutStatus for predicate IsProgressDeadlineExceeded.
+	PredicateProgressDeadlineExceeded: func(d *deploy) string {
+		return "Deployment exceeded its progress deadline"
+	},
+	// PredicateOlderReplicaActive refer to rolloutStatus for predicate IsOlderReplicaActive.
+	PredicateOlderReplicaActive: func(d *deploy) string {
+		if d.object.Spec.Replicas == nil {
+			return "Replica update in progress : some older replicas have been updated"
+		}
+		return fmt.Sprintf("Replica update in progress : %d out of %d new replicas have been updated",
+			d.object.Status.UpdatedReplicas, *d.object.Spec.Replicas)
+	},
+	// PredicateTerminationInProgress refer rolloutStatus for predicate IsTerminationInProgress.
+	PredicateTerminationInProgress: func(d *deploy) string {
+		return fmt.Sprintf("Replica termination in progress : %d old replicas are pending termination",
+			d.object.Status.Replicas-d.object.Status.UpdatedReplicas)
+	},
+	// PredicateUpdateInProgress refer to rolloutStatus for predicate IsUpdateInProgress.
+	PredicateUpdateInProgress: func(d *deploy) string {
+		return fmt.Sprintf("Replica update in progress : %d of %d updated replicas are available",
+			d.object.Status.AvailableReplicas, d.object.Status.UpdatedReplicas)
+	},
+	// PredicateNotSpecSynced refer to status rolloutStatus for predicate IsNotSyncSpec.
+	PredicateNotSpecSynced: func(d *deploy) string {
+		return "Deployment rollout in-progress : deployment rollout in-progress : waiting for deployment spec update to be observed"
+	},
+}
+
+// rolloutChecks contains a group of predicate it useses predicateName as key.
+var rolloutChecks = map[predicateName]predicate{
+	// PredicateProgressDeadlineExceeded refer to predicate IsProgressDeadlineExceeded.
+	PredicateProgressDeadlineExceeded: IsProgressDeadlineExceeded(),
+	// PredicateOlderReplicaActive refer to predicate IsOlderReplicaActive.
+	PredicateOlderReplicaActive: IsOlderReplicaActive(),
+	// PredicateTerminationInProgress refer to predicate IsTerminationInProgress.
+	PredicateTerminationInProgress: IsTerminationInProgress(),
+	// PredicateUpdateInProgress refer to predicate IsUpdationInProgress.
+	PredicateUpdateInProgress: IsUpdateInProgress(),
+	// PredicateNotSpecSynced refer to predicate IsSyncSpec.
+	PredicateNotSpecSynced: IsNotSyncSpec(),
+}
+
+// rolloutOutput struct contaons message and boolean value to show rolloutstatus
+type rolloutOutput struct {
```

Comment:
can we rename this to `rollout`

Reply:
Let us stick to this structure & methods from this structure.
Let us avoid builder for the timebeing since this structure consist of less fields
Let us make use of `type rolloutBuildOption func (*rollout)` & `func NewRollout(opts ...rolloutBuildOption) *rollout` if required.
If above seems to be over engineered then simple `func NewRollout(arg1 arg2) *rollout` will also suffice

Reply:
done